### PR TITLE
Widget Refactor with Context API

### PR DESCRIPTION
**Pull request by AndresVidalBerriel**
#### :link: Trello board reference:

- [Refactor del loop de mensajes del widget](https://trello.com/c/0mcNUjcz)

---

#### :pencil2: Description:

**This PR introduces the same changes as [this closed PR](https://github.com/NeoCoast/pis2019-chat-widget/pull/5) but uses Context API instead of Redux to manage global state**

The widget has been refactored due to the following reasons:

- Previous state management was inefficient, as every state was handled by top-level component `Widget`, causing unnecesary repeated renderization of the whole app.

- Message loss before connection is properly stablished (a chat id is assigned and the websocket is subscribed to the action cable channel) was prevented by disabling the input until the socket is ready. This compromises the usability of the app.

- Project structure was significantly different from it's dashboard counterpart, difficulting the coordination of the team and development standarization.

Changes introduced by this PR include:

- Encapsulation of API calls in a single ChatController class
- Configuration to use `Context API` and project reorganization so it's closer to dashboard project
- Implementation of a message queue that ensures that messages are sent once and when the connection is stablished
- Separation of business logic from presentation in some components
- Connection status management

---

#### :pushpin: Notes:
**To Do (in this task, going to be included in a different PR):**
- Message delivery confirmation (both message received by the server and message broadcasted in channel)
- Associated with last item, error handling when message delivery is not acknowledged or not broadcasted back by the server

**To Do (may be out of scope for this task):**
- Separation between presentation and logic on the other components
- Enhance error handling

### CODE REVIEW

File: src/context/messages/index.jsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+const initialState = {
+  queued: [],
+  shown: {},
+  sending: false,
+  newMessage: '',
+}
+
+const MessagesContext = createContext()
+
+// eslint-disable-next-line react/prop-types
```
Comment (agusit7) [inline src/context/messages/index.jsx]: Hey, proptypes are very useful to prevent type errors. Why did we disable them here?

Comment (AndresVidalBerriel) [reply src/context/messages/index.jsx]: Yeah, I didn't know right away which proptype rule to use there, so I disabled it and forgot to tidy it up later. I'm pushing a new commit with this correction.

**Comment by AndresVidalBerriel** [comment]
@agusit7

Well, with the new implementation I had to be more careful when designing the state's semantic (not just let redux handle it, as now we'll have to handle everything), so for me that's a clear improvement independently of the technology used. What I conceptually like more about the switch to Context API is the fact that we have three contexts (instead of one huge redux store), so the code is a bit more organized.

About the boilerplate, Context API is just a way to avoid prop drilling, it doesn't give an actual solution to state management, so design decisions about this have to be made from scratch. The states handled by `ChatContext` and `MessagesContext` are objects with several properties and `useReducer` hook provides a nice structure to define state transitions. Also, it matches the design pattern used in the dashboard, which I think is really important.

We could use `useState` for each one of these properties  and pass `[state, setState]` (or some wrapper functions) to the context, but we would end up spreading state transitions over the rest of the code. There are some ways to handle this better, but I don't see how them would require less boilerplate than `useReducer` hook. Another option would be to separate the Contexts even more and have more Context Providers to manage smaller state pieces so `useReducer` is not necessary. I think this would improve the performance, too, but also it would bulk the code up.

All of that considered, I think using Context API is ok for now. Switching to redux is not urgent, but it won't be hard if we decide to do it in the future, after a more careful investigation. Also, I might have over-engineered this while trying to implement something scalable, so simplifications are welcome.

Thank you very much for the review, and I hope I answered your questions!

**Comment by agusit7** [comment]
That sounds great to me! I just tested it and its working perfectly.

**Pull request merged**